src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,369 @@
+        internal readonly TypeSymbol RightConvertedType;
+#if DEBUG
+        internal abstract TreeDumperNode DumpCore();
+        internal string Dump() => TreeDumper.DumpCompact(DumpCore());
+#endif
+
+        private TupleBinaryOperatorInfo(TypeSymbol leftConvertedType, TypeSymbol rightConvertedType)
+        {
+            LeftConvertedType = leftConvertedType;
+            RightConvertedType = rightConvertedType;
+        }
+
+        internal class Single : TupleBinaryOperatorInfo
+        {
+            internal readonly BinaryOperatorKind Kind;
+            internal readonly MethodSymbol MethodSymbolOpt; // User-defined comparison operator, if applicable
+
+            // To convert the result of comparison to bool
+            internal readonly Conversion BoolConversion;
+            internal readonly UnaryOperatorSignature BoolOperator; // Information for op_true or op_false
+
+            internal Single(TypeSymbol leftConvertedType, TypeSymbol rightConvertedType, BinaryOperatorKind kind,
+                MethodSymbol methodSymbolOpt, Conversion boolConversion, UnaryOperatorSignature boolOperator) : base(leftConvertedType, rightConvertedType)
+            {
+                // If a user-defined comparison operator is present, then the operator kind must be user-defined
+                Debug.Assert(Kind.IsUserDefined() || (object)MethodSymbolOpt == null);
+
+                // If the return type of methodSymbolOpt is not bool, then there must be a boolConversion or boolOperator
+                Debug.Assert(BoolConversion != default || !Kind.IsDynamic() || (Kind.IsUserDefined() && MethodSymbolOpt.ReturnType.SpecialType == SpecialType.System_Boolean));
+                Debug.Assert((object)BoolOperator != null || !Kind.IsDynamic() || (Kind.IsUserDefined() && MethodSymbolOpt.ReturnType.SpecialType == SpecialType.System_Boolean));
+
+                Kind = kind;
+                MethodSymbolOpt = methodSymbolOpt;
+                BoolConversion = boolConversion;
+                BoolOperator = boolOperator;
+            }
+
+            internal override bool IsSingle()
+                => true;
+
+            public override string ToString()
+                => $"binaryOperatorKind: {Kind}";
+
+#if DEBUG
+            internal override TreeDumperNode DumpCore()
+            {
+                var sub = new List<TreeDumperNode>();
+                if ((object)MethodSymbolOpt != null)
+                {
+                    sub.Add(new TreeDumperNode("methodSymbolOpt", MethodSymbolOpt.ToDisplayString(), null));
+                }
+                sub.Add(new TreeDumperNode("leftConversion", LeftConvertedType.ToDisplayString(), null));
+                sub.Add(new TreeDumperNode("rightConversion", RightConvertedType.ToDisplayString(), null));
+
+                return new TreeDumperNode("nested", Kind, sub);
+            }
+#endif
+        }
+
+        internal class Nested : TupleBinaryOperatorInfo
```

Comment:
>Nested [](start = 23, length = 6)

I think "Nested" is a misnomer. It isn't that it is nested within something else, but that it *contains* multiple operations. I suggest calling it `Multiple`.